plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt

```diff
@@ -454,18 +455,24 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
   inner class MapClick : OnMapClickListener {
     /**
      * Called when the user clicks on the map view.
+     * Note that calling this method is blocking main thread until querying map for features is finished.
      *
      * @param point The projected map coordinate the user clicked on.
      * @return True if this click should be consumed and not passed further to other listeners registered afterwards,
      * false otherwise.
      */
     override fun onMapClick(point: Point): Boolean {
-      queryMapForFeatures(point) {
-        selectAnnotation(it)
-        clickListeners.forEach { listener ->
-          listener.onAnnotationClick(it)
+      val latch = CountDownLatch(1)
+      mapFeatureQueryDelegate.executeOnRenderThread {
+        queryMapForFeatures(point) {
+          selectAnnotation(it)
+          clickListeners.forEach { listener ->
+            listener.onAnnotationClick(it)
+          }
+          latch.countDown()
         }
       }
+      latch.await()
```

Comment:
do we actually want to set some timeout? Blocking main thread for a long time doesn't sound right

Reply:
Sounds like a nice action, I was also thinking about it!
@Chaoba could you please add timeout here and throw an exception if timeout is hit.
IMHO 3 seconds is enough to say for sure that some issue with QRF has happened

Reply:
Or 3 is too much, perhaps even 1 sec.. In general we query for 1 feature only, it should be almost immediate
cc @mapbox/maps-android for more thoughts here

Reply:
I have already set it to 2 sec when polishing codes locally. In fact, the query will return all the features at this point, but we will only use the first one. Not very sure whether 1 sec is enough when there are too many features, so 2 should be enough.

Reply:
It still does not sound very reliable.
I guess we should use let's say default of 3 sec and add an API to set this limit for the user as well.
I could imagine (in very-very theory) someone with 1 million annotations on the map and being OK to freeze UI for 10 seconds

Reply:
thinking it again, I think blocking the main thread for an unknown time is not really good, blocking the main thread for 5 seconds will result in an ANR. I think we should rethink the approach here.

Reply:
Approach could not be changed sadly. Clicking on a map is UI Android operation, UI is single-threaded in Android and in this particular case click should be returned only after QRF is finished. So, sadly, there is no solution if we talk about threading. This approach works pretty fine in real conditions, it may result in significant lags if annotation count is enormous but that also could be optimized on another level.

We could re-think this approach with coroutines but assuming we do not use them at all this solution seems like a good start to me

Reply:
could we at least try to add some filter or limit the queried layer to the QRF so that it wouldn't return large amount of features?

Reply:
We have the filter as a param while querying https://github.com/mapbox/mapbox-maps-android/pull/455/files/3f15b4e2c37c0c953af9193c60182baf0f66ab3b#diff-4439957e75cc10a9ecf0041db71889e34906978146ed8b09519c8e0cc7405a1fR641

Reply:
Made a test and it would cost about 200ms for 2000 annotations in the same point and  800ms for 20000 annotations. So 1s should cover most of the cases and I set 2s to make it more robust.